File: AlgorithmGoVersion/src/Algorithm/dynamic/lcs/LCS_Dynamic.go
package lcs

// LCS_Dynamic
// @author: 蔚蔚樱
// @date: 2022/8/218:46
// @author—Email: micromicrohard@outlook.com
// @description:
// @blogURL:
type LCS_Dynamic struct {
}

// 返回 字符串 source, pattern 的相似度【有几个字符相等】
func (lcs *LCS_Dynamic) GetCommonLength(source, pattern string) int {
	if source == "" || pattern == "" {
		return 0
	}
	numMatrix, _ := lcs.lcsOperation(source, pattern)
	return numMatrix[len(source)-1][len(pattern)-1]
}

func (lcs *LCS_Dynamic) GetSimilarityMethod(source, pattern string) string {
	if source == "" || pattern == "" {
		return ""
	}
	_, similarityMatrix := lcs.lcsOperation(source, pattern)
	return lcs.BackTrackLCS(similarityMatrix, pattern)
}

func (lcs *LCS_Dynamic) lcsOperation(source, pattern string) ([][]int, [][]string) {
	var (
		n1, n2           int        // 获取字符串长度,其中在矩阵中n1表示行数,n2表示列数
		numMatrix        [][]int    // 数值矩阵 存放字符匹配数量的最大值
		similarityMatrix [][]string // 标记矩阵 存放字符匹配标记
	)
	n1 = len(source)
	n2 = len(pattern)
	numMatrix = make([][]int, n1+1)
	similarityMatrix = make([][]string, n1+1)
	for i, _ := range numMatrix {
		numMatrix[i] = make([]int, n2+1)
		similarityMatrix[i] = make([]string, n2+1)
	}

	for i := 1; i <= n1; i++ {
		for j := 1; j <= n2; j++ {
			a := string(source[i-1])
			b := string(pattern[j-1])
			if a == b {
				numMatrix[i][j] = numMatrix[i-1][j-1] + 1
				similarityMatrix[i][j] = "match"
			} else if numMatrix[i-1][j] > numMatrix[i][j-1] {
				numMatrix[i][j] = numMatrix[i-1][j]
				similarityMatrix[i][j] = "up"
			} else {
				numMatrix[i][j] = numMatrix[i][j-1]
				similarityMatrix[i][j] = "left"
			}
		}
	}
	return numMatrix, similarityMatrix
}

func (lcs *LCS_Dynamic) BackTrackLCS(similarityMatrix [][]string, pattern string) string {
	var lcsStrArr []string // lsc串
	var lcsString string   // lsc串
	if similarityMatrix == nil || len(similarityMatrix) == 0 {
		return ""
	}
	// attention: i,j 并不是双重循环，不能直接 ++--
	row := len(similarityMatrix) - 1
	column := len(similarityMatrix[0]) - 1
	for row >= 0 && column >= 0 {
		if similarityMatrix[row][column] == "match" {
			lcsStrArr = append(lcsStrArr, string(pattern[column-1]))
			row--
			column--
		} else if similarityMatrix[row][column] == "up" {
			row--
		} else {
			column--
		}
	}
	// grammar string数组转string
	for i := len(lcsStrArr) - 1; i >= 0; i-- {
		lcsString += lcsStrArr[i]
	}
	return lcsString
}

func (lcs *LCS_Dynamic) InterfaceName() string {
	return "LCS"
}

File: AlgorithmJavaVersion/src/Algorithm/dynamic/lcs/LCS_Dynamic.java
package Algorithm.dynamic.lcs;

import Common.Utils.UTFactory;
import Top100.Dynamic;
import org.junit.Test;

/**
 * @author 蔚蔚樱
 * @version 1.0
 * @date 2018-7-19 Time：上午10:15:35
 * @author—Email micromicrohard@outlook.com
 * @description LCS算法，找出两个字符串最大匹配子串。
 */

public class LCS_Dynamic implements LCS, Dynamic {

    @Test // 验证功能：用于全量测试
    public void TestFunc() throws Exception {
        UTFactory.FullTest(this.getClass());
    }

    @Test // 调试功能 : 用于复现错误测试案例
    public void TestDoubleTrack() throws Exception {
        String input = "EmptyExpr  ############  NullExpr";
        String output = "EmptyExpr";
        UTFactory.DebugTest(this.getClass(), input, output);
    }

    int n1, n2;                // 获取字符串长度,其中在矩阵中n1表示行数,n2表示列数
    int[][] numMatrix;         // 数值矩阵 存放字符匹配数量的最大值
    char[][] similarityMatrix; // 标记矩阵 存放字符匹配标记
    String lcsString = null;   // lsc串

    public int getCommonLength(String source, String pattern) {
        lcsOperation(source, pattern);
        return numMatrix[n1][n2];
    }

    public String GetSimilarityMethod(String source, String pattern) {

        LCS_Dynamic ans = lcsOperation(source, pattern);
        if (ans == null) {
            return "";
        }
        //找寻匹配上的字符串
        StringBuilder sb = new StringBuilder();
        for (int i = n1, j = n2; i > 0 && j > 0; ) {
            if (similarityMatrix[i][j] == '↖') {
                sb.append(source.charAt(i - 1));
                i--;
                j--;
            } else if (similarityMatrix[i][j] == '↑') {
                i--;
            } else {
                j--;
            }
        }
        lcsString = sb.length() == 0 ? "" : sb.reverse().toString();
        return lcsString;
    }

    public LCS_Dynamic lcsOperation(String source, String pattern) {

        if (!check(source, pattern)) {
            return null;
        }
        n1 = source.length();
        n2 = pattern.length();
        numMatrix = new int[n1 + 1][n2 + 1];
        similarityMatrix = new char[n1 + 1][n2 + 1];

        //数值矩阵初始化（若未初始化,数值矩阵显示为0,标记矩阵显示为口）
        for (int i = 0; i < n1; i++) {
            numMatrix[i][0] = 0;
            similarityMatrix[i][0] = 0;
        }
        for (int j = 0; j < n2; j++) {
            numMatrix[0][j] = 0;
            similarityMatrix[0][j] = 0;
        }

        //矩阵匹配,从每一行（或列）开始匹配
        //初始化的部分不用考虑，匹配从1开始
        for (int row = 1; row <= n1; row++) {
            //遍历列
            for (int column = 1; column <= n2; column++) {
                numMatrix[row][column] = 0;//初始化数值矩阵
                similarityMatrix[row][column] = 0;//初始化标记矩阵

                if (source.charAt(row - 1) == pattern.charAt(column - 1)) {//字符串元素匹配相等
                    numMatrix[row][column] = numMatrix[row - 1][column - 1] + 1;//取矩阵左上角数值加一
                    similarityMatrix[row][column] = '↖';//匹配标记（方便后面寻找）
                } else {//字符串元素匹配不等
                    if (numMatrix[row - 1][column] > numMatrix[row][column - 1]) {
                        numMatrix[row][column] = numMatrix[row - 1][column];//取矩阵上一行数值
                        similarityMatrix[row][column] = '↑';//向上寻找标记
                    } else {
                        numMatrix[row][column] = numMatrix[row][column - 1];//取矩阵左边数值
                        similarityMatrix[row][column] = '←';//向左寻找标记
                    }
                }
            }
        }
        return this;
    }
}
